library/std/src/sys/sync/rwlock/queue.rs

```diff
@@ -470,108 +495,122 @@ impl RwLock {
-                // set, meaning `read_unlock` will realize that the lock is still
-                // exclusively locked and act accordingly.
-                let next = state.wrapping_byte_add(DOWNGRADE);
+                // Another thread already holds the queue lock. Tell it to wake up all waiters.
+                // If the other thread succeeds in waking up waiters before we release our lock, the
+                // effect will be just the same as if we had changed the state below.
+                // Otherwise, the `DOWNGRADED` bit will still be set, meaning that when this thread
+                // calls `read_unlock` later, it will realize that the lock is still exclusively
+                // locked and act accordingly.
+                let next = state.wrapping_byte_add(DOWNGRADED);
                 match self.state.compare_exchange_weak(state, next, Release, Relaxed) {
                     Ok(_) => return,
                     Err(new) => state = new,
                 }
             } else {
-                // Grab the whole queue.
+                // Grab the entire queue by swapping the `state` with a single reader.
                 let next = ptr::without_provenance_mut(SINGLE | LOCKED);
                 if let Err(new) = self.state.compare_exchange_weak(state, next, AcqRel, Relaxed) {
                     state = new;
                     continue;
                 }
 
-                // Wake up all waiters in FIFO order.
+                // SAFETY: We have full ownership of this queue now, so nobody else can modify it.
                 let tail = unsafe { add_backlinks_and_find_tail(to_node(state)) };
-                // SAFETY: `tail` was just computed, meaning the whole queue is
-                // linked. Since the queue was not locked and we are not in read
-                // mode, we have complete control over this queue.
-                return unsafe { complete_all(tail) };
+
+                // Wake up all waiters.
+                // SAFETY: `tail` was just computed, meaning the whole queue is linked.
+                unsafe { complete_all(tail) };
+
+                return;
             }
         }
     }
 
-    /// Unlocks the queue. Wakes up all threads if a downgrade was requested,
-    /// otherwise wakes up the next eligible thread(s) if the lock is unlocked.
+    /// Unlocks the queue. Wakes up all threads if a downgrade was requested, otherwise wakes up the
+    /// next eligible thread(s) if the lock is unlocked.
     ///
     /// # Safety
-    /// The queue lock must be held by the current thread.
+    ///
+    /// * The queue lock must be held by the current thread.
+    /// * There must be threads queued on the lock.
     unsafe fn unlock_queue(&self, mut state: State) {
         debug_assert_eq!(state.addr() & (QUEUED | QUEUE_LOCKED), QUEUED | QUEUE_LOCKED);
 
         loop {
+            // SAFETY: FIXME why exactly is this safe?
```

Comment:
I'm not 100% sure why this is safe given the safety requirement of `unlock_queue`.  Is it because the threads that are modifying the queue can only add backlinks and cannot modify specifically the tail?